Add org.jousse.vincent.Pomodorolm

vjousse opened the pull request:
<!-- ⚠️  The submission PR must be against the `new-pr` branch ⚠️  -->

### Please confirm your submission meets all the criteria

<!-- Please replace each `[ ]` with `[X]` when the step is complete -->

- [x] Please describe the application briefly.

Pomodorolm is a simple, good looking and configurable pomodoro tracker with tray icon. You can change its look and feel by using themes and you if none suits you, you can create your own theme.
All sessions durations are configurable and you can choose to play sounds too (ticking sound, end of session and so on). You can even display notifications if you want to.


- [x] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [x] I have read the and followed all the [Submission and licence requirements][reqs].
- [x] I have [built][build] and tested the submission locally.
- [x] I am an author/developer/upstream contributor of the project. If not, I contacted upstream developers about this submission. **Link:**


<!-- ⚠️  Don't modify anything below this line ⚠️  -->

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

bbhtt commented on org.jousse.vincent.Pomodorolm.yml:
```diff
@@ -0,0 +1,45 @@
+  - --own-name=org.kde.*
+  # Needed for the systray to work properly. libappindicator emits a dbus signal with the systray icon absolute path
+  # so flatpak and the host need to both have access to the icon for the systray to work properly
+  - --filesystem=xdg-data
```
this won't do, you should not ask for access to entire xdg-data. At best the icon could be in a subdirectory or in tmp folder.

bbhtt replied:
Another approach is that you just supply the icon name for tray: `org.jousse.vincent.Pomodorlm-tray.png`, appindicators can pick them up as Flatpak will export them and set `XDG_
DATA_DIRS`

vjousse replied:
Well the problem is that I'm using [Tauri](https://v2.tauri.app) and I don't have access directly to the appindicator. Sharing `xdg-data/tray-icon` between the two should do the trick I think.

bbhtt commented on org.jousse.vincent.Pomodorolm.yml:
```diff
@@ -0,0 +1,45 @@
+    buildsystem: simple
+    sources:
+      - type: file
+        url: https://github.com/vjousse/pomodorolm/releases/download/app-v0.1.8/pomodorolm_0.1.8_amd64.deb
+        sha256: ff1b5b6ed87489f446c852ef47e5bec1018347a371518bfe7411f932ecad88a4
```
why not build from source?

vjousse replied:
Building from the deb package is how the official Tauri doc recommends to it https://v2.tauri.app/distribute/flatpak/
Doing it in the flatpak manifest would require setting up rust and npm build tools. Is it something recommended?

vjousse replied:
Thanks the pointers, I'll give it a try. I suppose that it will be a good opportunity to update the tauri documentation :+1:

bbhtt commented on screenshots/pomodorolm.png:
screenshots should be upstream

bbhtt commented on org.jousse.vincent.Pomodorolm.yml:
```diff
@@ -0,0 +1,45 @@
+  - --talk-name=org.freedesktop.Notifications
+  - --talk-name=org.kde.StatusNotifierWatcher
+  # Required for the systray to work
+  - --own-name=org.kde.*
```
See https://docs.flathub.org/docs/for-app-authors/linter#finish-args-wildcard-kde-own-name

https://docs.flathub.org/docs/for-app-authors/linter#exceptions

hfiguiere commented on org.jousse.vincent.Pomodorolm.desktop:
```diff
@@ -0,0 +1,11 @@
+[Desktop Entry]
+Version=0.1.7
```
this version field is invalid. It's not the application version.

hfiguiere commented on org.jousse.vincent.Pomodorolm.yml:
```diff
@@ -0,0 +1,45 @@
+      - ar -x *.deb
+      - tar -xf data.tar.gz
+      - "install -Dm755 usr/bin/pomodorolm /app/bin/pomodorolm"
+      - mkdir -p /app/usr/lib
+      - cp -rf usr/lib/pomodorolm /app/usr/lib/
```
this is phishy. `/app/usr/lib` isn't a standard. it should be `/app/lib` because here the prefix is `/app` not `/usr`.

hfiguiere commented on org.jousse.vincent.Pomodorolm.metainfo.xml:
```diff
@@ -0,0 +1,36 @@
+  <description>
+    <p>
+      Pomodorolm is a simple, good looking and configurable pomodoro tracker with tray icon. You can change its look and feel by using themes and you if none suits you, you can create your own theme.
+
```
shouldn't there be another `<p>` tag here?

hfiguiere commented on org.jousse.vincent.Pomodorolm.metainfo.xml:
```diff
@@ -0,0 +1,36 @@
+
+  <releases>
+      <release version="0.1.8" date="2024-09-06">
+         <description></description>
```
```suggestion
```

hfiguiere commented on org.jousse.vincent.Pomodorolm.metainfo.xml:
```diff
@@ -0,0 +1,36 @@
+  <launchable type="desktop-id">org.jousse.vincent.Pomodorolm.desktop</launchable>
+  <screenshots>
+    <screenshot type="default">
+      <image>screenshots/pomodorolm.png</image>
```
this is invalid. It should be a publicly accessible URL.

hfiguiere commented on org.jousse.vincent.Pomodorolm.metainfo.xml:
```diff
@@ -0,0 +1,36 @@
+  <id>org.jousse.vincent.Pomodorolm</id>
+
+  <name>pomodorolm</name>
+  <summary>A simple, good looking and configurable pomodoro tracker with tray icon.</summary>
```
```suggestion
  <summary>A simple, good looking and configurable pomodoro tracker with tray icon</summary>
```

hfiguiere commented on org.jousse.vincent.Pomodorolm.metainfo.xml:
```diff
@@ -0,0 +1,36 @@
+
+  <description>
+    <p>
+      Pomodorolm is a simple, good looking and configurable pomodoro tracker with tray icon. You can change its look and feel by using themes and you if none suits you, you can create your own theme.
```
if only one knew what "pomodoro" is, one could better understand what this is.

vjousse commented on flathub.json:
Well, why not, but I wouldn't be able to test it locally. Is it a problem? The app is built using [tauri](https://v2.tauri.app/), not sure if it supports it.

bbhtt replied:
there are lots of tauri apps on Flathub that have aarch64 builds.

hfiguiere commented on org.jousse.vincent.Pomodorolm.desktop:
it's not in the .deb either?

hfiguiere commented on org.jousse.vincent.Pomodorolm.metainfo.xml:
it's not in the .deb either?

hfiguiere commented on org.jousse.vincent.Pomodorolm.png:
it's not in the .deb either?

vjousse commented:
I have addressed all of your concerns (at least I think so) and built the whole project from source inside the flatpak. I've also removed unneeded `finish-args`. Don't hesitate if there is something that I should do differently.

bbhtt commented on org.jousse.vincent.Pomodorolm.yml:
```diff
@@ -0,0 +1,83 @@
+  - --talk-name=org.kde.StatusNotifierWatcher
+  # Needed for the systray to work properly. libappindicator emits a dbus signal with the systray icon absolute path
+  # so flatpak and the host need to both have access to the icon for the systray to work properly
+  - --filesystem=xdg-data/tray-icon
```
I don't understand why this is still needed. `$XDG_DATA_HOME` inside the Flatpak sandbox is set to `~/.var/app/$FLATPAK_ID/data`.

So not having this permission will mean the tray icon goes in `~/.var/app/$FLATPAK_ID/data` which is a path accessible on host.

Why do you want it on host's $XDG_DATA_HOME which is usually ~/.local/share/tray-icon` ?

vjousse replied:
Indeed, it's not needed anymore, I've removed it

bbhtt commented on registry.dat:
What is the source of this binary file ?

vjousse replied:
This is a binary generated by the Elm compiler (https://github.com/elm/compiler/blob/2f6dd29258e880dbb7effd57a829a0470d8da48b/builder/src/Stuff.hs#L147) when installing packages locally. It looks like it's an index of the packages installed. The problem is that there is no way to generate this file without internet access, and my Haskell skills are not good enough to understand how the Elm compiler generate it.

Here is how I generate it locally:
```
rm -rf elm-stuff && ELM_HOME=tmp npx elm make src-elm/Main.elm
cp tmp/0.19.1/packages/registry.dat flatpak
rm -rf tmp
```

I clean the `elm` cache, I compile the app, required packages will be downloaded to `tmp` and the registry.dat corresponding to my dependencies will be generated. Without this file, the elm will try to connect to internet and fail to compile.

bbhtt replied:
Also what's the format of the file ?

vjousse replied:
Unfortunately it's not documented anywhere… I found some bash code that creates it from scratch https://github.com/robx/shelm/blob/master/shelm#L224
That't what I'm using inside the flatpak now to generate the `registry.dat`. So this is basically the list of all the available packages of the registry but binary encoded. After discussing with folks on the Elm Slack channel, it looks like it was coded this way for performance reasons.

bbhtt commented on org.jousse.vincent.Pomodorolm.yml:
```diff
@@ -0,0 +1,84 @@
+      # flatpak-elm-generator.py ../elm.json elm-sources.json
+      - elm-sources.json
+
+      - type: file
+        path: shelm
```
Let's use this as URL

```
- type: file
  url: https://raw.githubusercontent.com/robx/shelm/4417730f5847e6ccba1b19f1b25166471433d633/shelm
  sha256:  ....
```

Copying it may cause copyright issues.

vjousse replied:
Should be fixed (with an additional patch file)

bbhtt commented on shelm:
Remove this file from here

bbhtt commented:
bot, build org.jousse.vincent.Pomodorolm

Feel free to start the build as you fix issues.

vjousse commented:
bot, build org.jousse.vincent.Pomodorolm

vjousse commented:
bot, build org.jousse.vincent.Pomodorolm

hfiguiere commented on org.jousse.vincent.Pomodorolm.yml:
```diff
@@ -0,0 +1,88 @@
+    sources:
+      - type: git
+        url: https://github.com/vjousse/pomodorolm.git
+        branch: flathub
```
NEVER a branch.

vjousse commented:
bot, build org.jousse.vincent.Pomodorolm

hfiguiere commented on org.jousse.vincent.Pomodorolm.yml:
```diff
@@ -28,7 +28,7 @@ modules:
       - type: git
         url: https://github.com/vjousse/pomodorolm.git
-        branch: flathub
+        tag: flathub
```
lying doesn't help

![image](https://github.com/user-attachments/assets/f276c2ba-034e-41cb-8fb3-62bf2dcbad55)

hfiguiere replied:
```suggestion
```

vjousse replied:
What are you talking about and why are you so aggressive?
![image](https://github.com/user-attachments/assets/276fa517-748b-413b-b827-add09c1420e8)

Just trying to do my best here, no need to YELL at me in your previous comment nor to say that I'm a liar when I'm not.

hfiguiere commented:
```
npm error I am detecting that your computer (linux_arm64) may not be compatible with any
npm error of the official pre-built binaries.
```

exotic languages defeating the portability.

bbhtt commented:
elm doesn't seem to have linux aarch64 binaries available.

vjousse commented:
bot, build org.jousse.vincent.Pomodorolm

vjousse commented:
Fixed the `aarch64` build by using https://www.npmjs.com/package/@lydell/elm that provides additional binaries.

vjousse commented:
bot, build org.jousse.vincent.Pomodorolm
